### PR TITLE
fix(compactor): empty list panic

### PR DESCRIPTION
**Pull request by Joe-Blount**
helps: https://github.com/influxdata/idpe/issues/17902

This PR addresses a scenario where we can hit a panic because a list of files to operate on is empty.

The criteria for this panic is:
- we have enough small files in a CompactRange to trigger ManySmallFiles,
- after splitting up files in the CompactRange into branches, one of the branches must be comprised entirely of files big enough that ManySmallFiles decides to skip them.

The first commit addresses the panic by making file_classification_for_many_files handle the case where every file in a branch gets skipped because its not 'small'.  A test is added to force this situation.

However, in the first commit, that new test results in the partition being put on the skip table, which is unnecessary, and the information indicating why its skipped is inaccurate.  There is a place in the code that assumes if we decide to work on no files, the only reason that could happen is because its huge data for a single ns of time, which we can't work on.  So that is the misleading explanation why the partition went on the skip table.

So the second commit removes that assumption that this situation is always a single ns situation.  To do so, the post classification filter needs to see the other files (which we decided not to work on), and check them for size and time range.
With the second commit, the new test added in the first commit compacts nicely - and all the single-timestamp test cases still work, without any change in their insta output.

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: compactor/src/components/file_classifier/split_based.rs [hunk 1]
```diff
@@ -295,12 +295,16 @@ fn file_classification_for_many_files(
         partition
     );
 
+    // We may have skipped everything in this branch.
+    let split_or_compact: FilesToSplitOrCompact = if files_to_compact.is_empty() {
+        FilesToSplitOrCompact::None(NoneReason::FilesSkipped)
+    } else {
+        FilesToSplitOrCompact::Compact(files_to_compact, CompactReason::ManySmallFiles)
```
Comment (savage-engineer) [inline compactor/src/components/file_classifier/split_based.rs:302]: Drive-by suggestion 🏎️ 

You could perhaps achieve this in a reusable way by implementing a `fn from(files_to_compact: Vec<ParquetFile>, reason: CompactReason) -> FilesToSplitOrCompact`. There may be reasons not to that I'm unaware of, but I thought I'd drop a comment!

File: compactor/src/components/post_classification_partition_filter/possible_progress.rs [hunk 2]
```diff
@@ -33,21 +34,29 @@ impl PostClassificationPartitionFilter for PossibleProgressFilter {
...
-                ),
-            )
-            .into())
+            for f in files_to_keep {
+                if f.file_size_bytes >= self.max_parquet_bytes as i64 && f.min_time == f.max_time {
+                    return Err(SimpleError::new(
+                        ErrorKind::OutOfMemory,
+                        format!(
+                            "partition {} has overlapped files that exceed max compact size limit {}. \
+                            This may happen if a large amount of data has the same timestamp",
```
Comment (wiedld) [inline compactor/src/components/post_classification_partition_filter/possible_progress.rs:50]: Requested minor change:
With the second conditional of `f.min_time == f.max_time`, we are only skipping if the data is too big to split (because within 1 ns). Could we update the error msg (which I believe gets passed to the skip table) to be more explicit that it's unsplitable?

Or at least something that makes it evident (when we query the catalog skip table) how this is 1ns resolution (not splitable) -- versus some other dense data distribution error.

**Review by wiedld — APPROVED** [review]
One small requested change to the logged reason to skip -- in order to be nice to our future selves.

Otherwise, looks good!
